// steam-market/index.js
var request = require('request');
var querystring = require('querystring');

var marketURL = 'http://steamcommunity.com/market/';
//priceoverview/?currency=5&appid=730&market_hash_name=StatTrak%E2%84%A2%20P250%20%7C%20Steel%20Disruption%20%28Factory%20New%29

var SteamMarket = function () {
	// Глобальное хранилище кук
	this._jar = request.jar();
	this._request = request.defaults({
		jar: this._jar
	});
}

SteamMarket.prototype.priceOverview = function (currency, appId, itemName, callback) {
	callAPI.bind(this)({
		method: 'priceoverview/',
		params: {
			currency: currency,
			appid: appId,
			market_hash_name: itemName
		},
		callback: callback
	});
}

function callAPI(options) {
	var httpMethod = options.post ? 'post' : 'get';

	var params = {
		uri: marketURL + options.method + '?' + querystring.stringify(options.params),
		json: true
	};

	if (options.post) {
		params.form = options.params;
	}

	request[httpMethod](params, function (error, response, body) {
		if (error || response.statusCode !== 200) {
			if (typeof options.callback === 'function') {
				options.callback(error || new Error(response.statusCode));
			}
			return;
		}
		if (!body || typeof body !== 'object') {
			if (typeof options.callback === 'function') {
				options.callback(new Error('Invalid Response'));
			}
			return;
		}
		if (typeof options.callback === 'function') {
			options.callback(null, body);
		}
	}.bind(this));
}

module.exports = new SteamMarket();
